README
virtualmail-pop3d (or vm-pop3d)

vm-pop3d is a POP3 server.

It supports alternative password files and mail spool directories;
it can be used for setting up virtual email accounts -- mailboxes
without real Unix owners for each. This will allow you to have
multiple email accounts with the same name on one system.

The POP3 protocol (by itself) is not secure. (Maybe vm-pop3d can
help by allowing you to use non-real system users.)

vm-pop3d is based on Jakob Kaivo's gnu-pop3d (formerly called ids-pop3d).
I developed ten patches for gnu-pop3d version 0.9.8; accordingly,
the first release of vm-pop3d is version 1.0.8. The code for vm-pop3d
version 1.0.8 should be identical to a virtual mail version 1.0-patched
gnu-pop3d 0.9.8.

vm-pop3d provides UIDL support. This means it can return unique-id's
for each message and the POP3 client can know which messages it has
retrieved previously. This can save a lot of time and bandwidth. On
related note, POP3 is not a good protocol for storing lots of old
mail on a server; all the mail is stored in one huge mailbox, causing
the POP3 servers to be very sluggish (although vm-pop3d is a lot faster
than most other POP3 servers). Do consider deleting the mail often.

This program is licensed under the terms of the GNU General
Public License (GPL). See the file COPYING for this license.
I am sorry the license is so restrictive; if you need a more free
license which you can freely patch and use in proprietry products,
please let me know.

See the AUTHORS file for information about the authors of this software.

Please see the file INSTALL for installation instructions. The INSTALL
document also has information on configuring for virtual users usage.

For "Frequently asked questions" read the FAQ.

A history of changes of vm-pop3d back through the initial
virtual mail patches for gnu-pop3d is available in the CHANGES file.

An example vm-pop3d.spec file is provided to help build an RPM for
this package. (If you make a vm-pop3d.spec file, RPMs, debs, or
other packages, please share.)

The RFC/ directory contains several RFCs related to POP3.
(The following RFCs are obsolete or replaced:
RFC1081, RFC1225, RFC1460, RFC1725 and RFC2095.)

http://www.reedmedia.net/software/virtualmail-pop3d/

Professional, commercial, technical support is available to
install, implement, and configure this virtual email hosting solution.

This is a freely-available open source solution for virtual
email hosting.

If you use vm-pop3d, please consider signing up for the
vm-pop3d-announce mailing list (to receive messages about
new versions, bug fixes, and security updates).

Mailing Lists                  

 vm-pop3d-announce - for announcements of new releases, bug fixes,
   or security problems. (No discussion/no posting).
 vm-pop3d-users - for discussion on installing and using vm-pop3d.
 vm-pop3d-devel - for discussion on code development of vm-pop3d.

To sign up for any of these mailing lists, send an email to
majordomo@lists.kmv.ru and in the body of the message put:
   subscribe listname 



CHANGES
CHANGES / HISTORY

17/Jan/2002 1.1.6
	Bug Fix: Don't freopen() mailbox in quit.c; it is now open
	read/write in user.c (and apop.c). (This freopen caused 
	fcntl(2) lock to be lost and mailbox could be corrupted if
	LDA didn't use lockfile checking.)

	Add getopt* source files to DISTFILES in Makefile.in. (Thanks
	to Andrey J. Melnikoff.) (They were removed from the SRCS list
	in the Makefile.in in the 1.1.5 version.)

21/Dec/2001 1.1.6
	More informative message (and free lockfile) if having problem
	when deleting .lock file. Also, unlock mbox first (extra.c).

19/Dec/2001 1.1.6
	Got rid of realloc() in pop3_retr (retr.c) and pop3_top (top.c);
	this was supposed to be removed with fixes on 11/Oct/2001 and
	28/Nov/2001.

	Only allow "From " to be delimiter of mbox message if it
	is the first line or has a blank line before it. (Some MDAs
	escape it with a ">" so it doesn't matter.) Fixes in
	extra.c (pop3_getsizes), retr.c (pop3_retr), top.c (pop3_top),
	and quit.c (pop3_quit). (Reported by Andrey J. Melnikoff
	and Jim Pazarena.)

18/Dec/2001 1.1.6
	Fix configure script (configure.in) so it explains that
	PAM and virtual-based support are enabled by default.

04/Dec/2001 1.1.6
	Documentation: Updated and clarified INSTALL document.

29/Nov/2001 1.1.6
	Documentation: Explained "host name not found" in FAQ.
	
29/Nov/2001 1.1.5
	Install manual page; version 1.1.1 accidently removed
	manual page installation (Makefile.in).

	Got rid of 600 second timeout limitation (vm-pop3d.c, vm-pop3d.8).

	Changed --help usage message (extra.c).

	Use system's getopt_long() if available; also, improved PAM
	checks (configure.in).

	Documentation: Added comments about PAM to INSTALL doc.

28/Nov/2001 1.1.5
	Fixed never-released bug introduced on 11/Oct/2001: if a dot
	(period) is at 80th character don't escape with another dot (retr.c).

	Fixed memory allocation in virtual.c, where it could be abused
	to use up some memory. Now the returned passwd data is just
	static. (Thanks to Jeff Davis.)

	Bug Fix: Removed freopen() calls in pop3_top (top.c), pop3_retr
	(retr.c), uidl (uidl.c) and pop3_getsizes (extra.c); they were not
	needed (and they could cause corrupted mailbox because the shared
	lock is lost). And made pop3_getsizes() rewind to beginning of file.

	Minor change in pop3_retr() -- don't fgets() one more line if
	done (retr.c).

	Bug Fix: Changed top.c so it doesn't try to place entire long
	line (up to newline) into memory before using. (See 11/Oct/2001.)

	Changed VIRTUAL_MAILDIR to VIRTUAL_MAILPATH so there won't be
	any confusion on whether it is maildir-type mailboxes (vm-pop3d.h,
	user.c). (At this time, it is only Berkeley-style mbox format.)

	Portability: Added ./configure directive to disable PAM
	(configure.in). If available, PAM is still enabled by default. If
	libdl exists, then use it (because some old versions of libpam
	didn't link to it). Changed cpp definitions to just be USE_PAM
	(config.h.in, vm-pop3d.h, user.c).

	Added PAM debugging message if PAM error (user.c).

	Fix: The syslogging is now reopened if there is a PAM error (user.c).
	(Problem reported by Ed Thomson.) Got rid of PAM_ERROR macro.

24/Nov/2001 1.1.5
	Documentation: Further explained "Bad login" in FAQ.

30/Oct/2001 1.1.5
	Added some debugging to pop3_quit() to report on number of
	messages to delete and total messages.

	Don't delete all messages in pop3_quit() if number of messages
	is zero and number of messages selected to be deleted is zero
	(quit.c). This issue occurred with a corrupted (or non-standard
	mailbox) as reported by Iliya Peregoudov; the mailbox was missing
	"From " message delimiter.

12/Oct/2001 1.1.5
	Documentation: Added FAQ about using MD5 passwords.
	(Information from Istvan Karaszi.)

11/Oct/2001 1.1.5
	Fixed copyright notices; some files had second copyright.
	(As far as it is known, original code has original author's
	copyright.)

	Bug Fix: Changed retr.c so it doesn't try to place entire long
	line (up to newline) into memory before printing. This worked,
	but is very slow and maybe cause denial of service problems. In
	addition, it was reported by zas of norz.org to crash some
	vm-pop3d services. This realloc() loop also reported by Andrey J.
	Melnikoff in July.

04/Oct/2001 1.1.5
	Documentation: Added FAQ about "Socket closed".

14/Sep/2001 1.1.5
	Made virtual.c work with password files with more fields
	(but it doesn't store the additional fields). This will
	allow it to work with standard passwd(5) format. (Idea from
	Tomi Hakala and others.)

10/Jul/2001 1.1.5
	Changed gnu-pop3d reference -- PAM_gnupop3d_conv to
	PAM_vmpop3d_conv (user.c).

	Documentation: Added FAQ about rejecting passwords ("-ERR Bad
	login") to remind to setup PAM configuration, if applicable.

26/Jun/2001 1.1.5
	A few debug logging were never done (vm-pop3d.c).

13/Jun/2001 1.1.5
	Bug Fix: --disable-virtual wouldn't compile; connect IP
	logging code added for 1.1.4 code used a variable that
	was previously just used for USE_VIRTUAL (vm-pop3d.c). 

11/Jun/2001 1.1.4
	At startup, make sure UID used for virtual exists
	(vm-pop3d.c).

	Documentation: Explained --user and --group options in
	manual page and INSTALL document; added FAQ entry about
	running as non-root.

	Feature: Added command-line option (-u or --user) to set
	user UID for setuid for virtual users (vm-pop3d.c and
	virtual.c).

	Feature: Added command-line option (-g or --group) to set
	GID for setgid (vm-pop3d.c).

	pop3_abquit now logs if fdopen() doesn't work for output
	stream (extra.c).

	If setuid didn't work then don't allow log in and log problem
	(user.c).

	Bug Fix: Creation of lockfile is checked. Before, if not
	created, it would seg fault at end of session. (Thanks
	to Rok Papez.)

	Bug Fix: When concurrent children exited or stopped at same time,
	the correct amount of new children wouldn't start. Status is
	collected for waitpid. SIGCHLD signal is set to SIG_DFL (the

md5.c
/* md5.c - Functions to compute MD5 message digest of files or memory blocks
   according to the definition of MD5 in RFC 1321 from April 1992.
   Copyright (C) 1995, 1996 Free Software Foundation, Inc.
   NOTE: The canonical source of this file is maintained with the GNU C
   Library.  Bugs can be reported to bug-glibc@prep.ai.mit.edu.

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 2, or (at your option) any
   later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

/* Written by Ulrich Drepper <drepper@gnu.ai.mit.edu>, 1995.  */

#include <sys/types.h>

#if 0				/* Get rid of implicit bcopy declaration
				 * warnings */
#if STDC_HEADERS || defined _LIBC
#include <stdlib.h>
#endif

#include <string.h>

#if 0
#else
#ifndef HAVE_MEMCPY
#define memcpy(d, s, n) bcopy ((s), (d), (n))
#endif
#endif
#endif

#include "md5.h"

#ifdef _LIBC
#include <endian.h>
#if __BYTE_ORDER == __BIG_ENDIAN
#define WORDS_BIGENDIAN 1
#endif
#endif

#ifdef WORDS_BIGENDIAN
#define SWAP(n)							\
    (((n) << 24) | (((n) & 0xff00) << 8) | (((n) >> 8) & 0xff00) | ((n) >> 24))
#else
#define SWAP(n) (n)
#endif


/* This array contains the bytes used to pad the buffer to the next
   64-byte boundary.  (RFC 1321, 3.1: Step 1)  */
static const unsigned char fillbuf[64] =
{0x80, 0 /* , 0, 0, ...  */ };


/* Initialize structure containing state of computation.
   (RFC 1321, 3.3: Step 3)  */
void
md5_init_ctx(ctx)
struct md5_ctx *ctx;
{
  ctx->A = 0x67452301;
  ctx->B = 0xefcdab89;
  ctx->C = 0x98badcfe;
  ctx->D = 0x10325476;

  ctx->total[0] = ctx->total[1] = 0;
  ctx->buflen = 0;
}

/* Put result from CTX in first 16 bytes following RESBUF.  The result
   must be in little endian byte order.

   IMPORTANT: On some systems it is required that RESBUF is correctly
   aligned for a 32 bits value.  */
void *
md5_read_ctx(ctx, resbuf)
const struct md5_ctx *ctx;
void *resbuf;
{
  ((md5_uint32 *) resbuf)[0] = SWAP(ctx->A);
  ((md5_uint32 *) resbuf)[1] = SWAP(ctx->B);
  ((md5_uint32 *) resbuf)[2] = SWAP(ctx->C);
  ((md5_uint32 *) resbuf)[3] = SWAP(ctx->D);

  return resbuf;
}

/* Process the remaining bytes in the internal buffer and the usual
   prolog according to the standard and write the result to RESBUF.

   IMPORTANT: On some systems it is required that RESBUF is correctly
   aligned for a 32 bits value.  */
void *
md5_finish_ctx(ctx, resbuf)
struct md5_ctx *ctx;
void *resbuf;
{
  /* Take yet unprocessed bytes into account.  */
  md5_uint32 bytes = ctx->buflen;
  size_t pad;

  /* Now count remaining bytes.  */
  ctx->total[0] += bytes;
  if (ctx->total[0] < bytes)
    ++ctx->total[1];

  pad = bytes >= 56 ? 64 + 56 - bytes : 56 - bytes;
  memcpy(&ctx->buffer[bytes], fillbuf, pad);

  /* Put the 64-bit file length in *bits* at the end of the buffer.  */
  *(md5_uint32 *) & ctx->buffer[bytes + pad] = SWAP(ctx->total[0] << 3);
  *(md5_uint32 *) & ctx->buffer[bytes + pad + 4] = SWAP((ctx->total[1] << 3) |
						  (ctx->total[0] >> 29));

  /* Process last bytes.  */
  md5_process_block(ctx->buffer, bytes + pad + 8, ctx);

  return md5_read_ctx(ctx, resbuf);
}

/* Compute MD5 message digest for bytes read from STREAM.  The
   resulting message digest number will be written into the 16 bytes
   beginning at RESBLOCK.  */
int
md5_stream(stream, resblock)
FILE *stream;
void *resblock;
{
  /* Important: BLOCKSIZE must be a multiple of 64.  */
#define BLOCKSIZE 4096
  struct md5_ctx ctx;
  char buffer[BLOCKSIZE + 72];
  size_t sum;

  /* Initialize the computation context.  */
  md5_init_ctx(&ctx);

  /* Iterate over full file contents.  */
  while (1) {
    /* We read the file in blocks of BLOCKSIZE bytes.  One call of the
     * computation function processes the whole buffer so that with the
     * next round of the loop another block can be read.  */
    size_t n;

    sum = 0;

    /* Read block.  Take care for partial reads.  */
    do {
      n = fread(buffer + sum, 1, BLOCKSIZE - sum, stream);

      sum += n;
    }
    while (sum < BLOCKSIZE && n != 0);
    if (n == 0 && ferror(stream))
      return 1;

    /* If end of file is reached, end the loop.  */
    if (n == 0)
      break;

    /* Process buffer with BLOCKSIZE bytes.  Note that BLOCKSIZE % 64 == 0 */
    md5_process_block(buffer, BLOCKSIZE, &ctx);
  }

  /* Add the last bytes if necessary.  */
  if (sum > 0)
    md5_process_bytes(buffer, sum, &ctx);

  /* Construct result in desired memory.  */
  md5_finish_ctx(&ctx, resblock);
  return 0;
}

/* Compute MD5 message digest for LEN bytes beginning at BUFFER.  The
   result is always in little endian byte order, so that a byte-wise
   output yields to the wanted ASCII representation of the message
   digest.  */
void *
md5_buffer(buffer, len, resblock)
const char *buffer;
size_t len;
void *resblock;
{
  struct md5_ctx ctx;

  /* Initialize the computation context.  */
  md5_init_ctx(&ctx);

  /* Process whole buffer but last len % 64 bytes.  */
  md5_process_bytes(buffer, len, &ctx);

  /* Put result in desired memory area.  */
  return md5_finish_ctx(&ctx, resblock);
}


void
md5_process_bytes(buffer, len, ctx)
const void *buffer;
size_t len;
struct md5_ctx *ctx;
{
  /* When we already have some bits in our internal buffer concatenate
   * both inputs first.  */
  if (ctx->buflen != 0) {
    size_t left_over = ctx->buflen;
    size_t add = 128 - left_over > len ? len : 128 - left_over;

    memcpy(&ctx->buffer[left_over], buffer, add);
    ctx->buflen += add;

    if (left_over + add > 64) {
      md5_process_block(ctx->buffer, (left_over + add) & ~63, ctx);
      /* The regions in the following copy operation cannot overlap.  */
      memcpy(ctx->buffer, &ctx->buffer[(left_over + add) & ~63],
	     (left_over + add) & 63);
      ctx->buflen = (left_over + add) & 63;
    }
    buffer = (const char *) buffer + add;
    len -= add;
  }
  /* Process available complete blocks.  */
  if (len > 64) {
    md5_process_block(buffer, len & ~63, ctx);
    buffer = (const char *) buffer + (len & ~63);
    len &= 63;
  }
  /* Move remaining bytes in internal buffer.  */
  if (len > 0) {
    memcpy(ctx->buffer, buffer, len);
    ctx->buflen = len;
  }
}


/* These are the four functions used in the four steps of the MD5 algorithm
   and defined in the RFC 1321.  The first function is a little bit optimized
   (as found in Colin Plumbs public domain implementation).  */
/* #define FF(b, c, d) ((b & c) | (~b & d)) */
#define FF(b, c, d) (d ^ (b & (c ^ d)))
#define FG(b, c, d) FF (d, b, c)
#define FH(b, c, d) (b ^ c ^ d)
#define FI(b, c, d) (c ^ (b | ~d))

/* Process LEN bytes of BUFFER, accumulating context into CTX.
   It is assumed that LEN % 64 == 0.  */

void
md5_process_block(buffer, len, ctx)
const void *buffer;
size_t len;
struct md5_ctx *ctx;
{
  md5_uint32 correct_words[16];
  const md5_uint32 *words = buffer;
  size_t nwords = len / sizeof(md5_uint32);
  const md5_uint32 *endp = words + nwords;
  md5_uint32 A = ctx->A;
  md5_uint32 B = ctx->B;
  md5_uint32 C = ctx->C;
  md5_uint32 D = ctx->D;

  /* First increment the byte count.  RFC 1321 specifies the possible
   * length of the file up to 2^64 bits.  Here we only compute the number
   * of bytes.  Do a double word increment.  */
  ctx->total[0] += len;
  if (ctx->total[0] < len)
    ++ctx->total[1];

  /* Process all bytes in the buffer with 64 bytes in each round of the
   * loop.  */
  while (words < endp) {
    md5_uint32 *cwp = correct_words;
    md5_uint32 A_save = A;
    md5_uint32 B_save = B;
    md5_uint32 C_save = C;
    md5_uint32 D_save = D;

    /* First round: using the given function, the context and a constant
     * the next context is computed.  Because the algorithms processing
     * unit is a 32-bit word and it is determined to work on words in
     * little endian byte order we perhaps have to change the byte order
     * before the computation.  To reduce the work for the next steps we
     * store the swapped words in the array CORRECT_WORDS.  */

#define OP(a, b, c, d, s, T)						\
      do								\
        {								\
	  a += FF (b, c, d) + (*cwp++ = SWAP (*words)) + T;		\
	  ++words;							\
	  CYCLIC (a, s);						\
	  a += b;							\
        }								\
      while (0)

    /* It is unfortunate that C does not provide an operator for cyclic
     * rotation.  Hope the C compiler is smart enough.  */
#define CYCLIC(w, s) (w = (w << s) | (w >> (32 - s)))

    /* Before we start, one word to the strange constants. They are
     * defined in RFC 1321 as
     * 
     * T[i] = (int) (4294967296.0 * fabs (sin (i))), i=1..64 */

    /* Round 1.  */
    OP(A, B, C, D, 7, 0xd76aa478);
    OP(D, A, B, C, 12, 0xe8c7b756);
    OP(C, D, A, B, 17, 0x242070db);
    OP(B, C, D, A, 22, 0xc1bdceee);
    OP(A, B, C, D, 7, 0xf57c0faf);
    OP(D, A, B, C, 12, 0x4787c62a);
    OP(C, D, A, B, 17, 0xa8304613);
    OP(B, C, D, A, 22, 0xfd469501);
    OP(A, B, C, D, 7, 0x698098d8);
    OP(D, A, B, C, 12, 0x8b44f7af);
    OP(C, D, A, B, 17, 0xffff5bb1);
    OP(B, C, D, A, 22, 0x895cd7be);
    OP(A, B, C, D, 7, 0x6b901122);
    OP(D, A, B, C, 12, 0xfd987193);
    OP(C, D, A, B, 17, 0xa679438e);
    OP(B, C, D, A, 22, 0x49b40821);

    /* For the second to fourth round we have the possibly swapped words
     * in CORRECT_WORDS.  Redefine the macro to take an additional first
     * argument specifying the function to use.  */
#undef OP
#define OP(f, a, b, c, d, k, s, T)					\
      do 								\
	{								\
	  a += f (b, c, d) + correct_words[k] + T;			\
	  CYCLIC (a, s);						\
	  a += b;							\
	}								\
      while (0)

    /* Round 2.  */
    OP(FG, A, B, C, D, 1, 5, 0xf61e2562);
    OP(FG, D, A, B, C, 6, 9, 0xc040b340);
    OP(FG, C, D, A, B, 11, 14, 0x265e5a51);
    OP(FG, B, C, D, A, 0, 20, 0xe9b6c7aa);
    OP(FG, A, B, C, D, 5, 5, 0xd62f105d);
    OP(FG, D, A, B, C, 10, 9, 0x02441453);
    OP(FG, C, D, A, B, 15, 14, 0xd8a1e681);
    OP(FG, B, C, D, A, 4, 20, 0xe7d3fbc8);
    OP(FG, A, B, C, D, 9, 5, 0x21e1cde6);
    OP(FG, D, A, B, C, 14, 9, 0xc33707d6);
    OP(FG, C, D, A, B, 3, 14, 0xf4d50d87);
    OP(FG, B, C, D, A, 8, 20, 0x455a14ed);
    OP(FG, A, B, C, D, 13, 5, 0xa9e3e905);
    OP(FG, D, A, B, C, 2, 9, 0xfcefa3f8);
    OP(FG, C, D, A, B, 7, 14, 0x676f02d9);
    OP(FG, B, C, D, A, 12, 20, 0x8d2a4c8a);

    /* Round 3.  */
    OP(FH, A, B, C, D, 5, 4, 0xfffa3942);
    OP(FH, D, A, B, C, 8, 11, 0x8771f681);
    OP(FH, C, D, A, B, 11, 16, 0x6d9d6122);
    OP(FH, B, C, D, A, 14, 23, 0xfde5380c);
    OP(FH, A, B, C, D, 1, 4, 0xa4beea44);
    OP(FH, D, A, B, C, 4, 11, 0x4bdecfa9);
    OP(FH, C, D, A, B, 7, 16, 0xf6bb4b60);
    OP(FH, B, C, D, A, 10, 23, 0xbebfbc70);
    OP(FH, A, B, C, D, 13, 4, 0x289b7ec6);
    OP(FH, D, A, B, C, 0, 11, 0xeaa127fa);
    OP(FH, C, D, A, B, 3, 16, 0xd4ef3085);
    OP(FH, B, C, D, A, 6, 23, 0x04881d05);
    OP(FH, A, B, C, D, 9, 4, 0xd9d4d039);
    OP(FH, D, A, B, C, 12, 11, 0xe6db99e5);
    OP(FH, C, D, A, B, 15, 16, 0x1fa27cf8);
    OP(FH, B, C, D, A, 2, 23, 0xc4ac5665);

    /* Round 4.  */
    OP(FI, A, B, C, D, 0, 6, 0xf4292244);
    OP(FI, D, A, B, C, 7, 10, 0x432aff97);
    OP(FI, C, D, A, B, 14, 15, 0xab9423a7);
    OP(FI, B, C, D, A, 5, 21, 0xfc93a039);
    OP(FI, A, B, C, D, 12, 6, 0x655b59c3);
    OP(FI, D, A, B, C, 3, 10, 0x8f0ccc92);
    OP(FI, C, D, A, B, 10, 15, 0xffeff47d);
    OP(FI, B, C, D, A, 1, 21, 0x85845dd1);
    OP(FI, A, B, C, D, 8, 6, 0x6fa87e4f);
    OP(FI, D, A, B, C, 15, 10, 0xfe2ce6e0);
    OP(FI, C, D, A, B, 6, 15, 0xa3014314);
    OP(FI, B, C, D, A, 13, 21, 0x4e0811a1);
    OP(FI, A, B, C, D, 4, 6, 0xf7537e82);
    OP(FI, D, A, B, C, 11, 10, 0xbd3af235);
    OP(FI, C, D, A, B, 2, 15, 0x2ad7d2bb);
    OP(FI, B, C, D, A, 9, 21, 0xeb86d391);

    /* Add the starting values of the context.  */
    A += A_save;
    B += B_save;
    C += C_save;
    D += D_save;
  }

  /* Put checksum in context given as argument.  */
  ctx->A = A;
  ctx->B = B;
  ctx->C = C;
  ctx->D = D;
}
